### PR TITLE
Various improvements and fixes

### LINKED ISSUE #125: Some issues with path handling
This extends #15.

- The default output directory should relative to `$PWD` and not relative to the path of the .sby file, as it is now.
- Setting the work directory should work even when having multiple tasks. In that case, one sub-directory shall be created per task.
- We may want input paths (SystemVerilog files) to be relative to `$PWD`, or relative to the .sby file, or relative to the working/output directory. A CLI flag to set this would be appropriate.

### PR DESCRIPTION
**Pull request by piegamesde**
Closes #125. I've discovered that SymbiYosys does environment variable expansion within its paths. This can be trivially used to make the input paths relative to something else.

I've also changed a few print statements to make them more helpful.

### CODE REVIEW

**Comment by piegamesde** [comment]
I went ahead and changed all strings that used the old `.format()` syntax to f-strings while I was there. Also improved the error message. Some more refactoring change may be coming.

File: sbysrc/sby_engine_smtbmc.py [hunk 1]
```diff
@@ -68,7 +68,7 @@ def run(mode, job, engine_idx, engine):
         elif o == "--seed":
             random_seed = a
         else:
-            job.error("Invalid smtbmc options {}.".format(o))
+            job.error(f"Invalid smtbmc options {p}.")
```
Comment (piegamesde) [inline sbysrc/sby_engine_smtbmc.py:71]: Should be `job.error(f"Invalid smtbmc options {o}.")`

File: sbysrc/sby_engine_btor.py [hunk 2]
```diff
@@ -198,20 +202,23 @@ def exit_callback(retcode):
 
         if common_state.solver_status == "unsat":
             if common_state.expected_cex == 1:
-                with open("{}/engine_{}/trace.wit".format(job.workdir, engine_idx), "w") as wit_file:
+                with open(f"""{job.workdir}/engine_{engine_idx}/trace.wit""", "w") as wit_file:
```
Comment (piegamesde) [inline sbysrc/sby_engine_btor.py:205]: I don't think the triple quotes are needed here.

File: sbysrc/sby_engine_smtbmc.py [hunk 3]
```diff
@@ -177,37 +182,37 @@ def output_callback(line):
...
         if task_status is None:
-            job.error("engine_{}: Engine terminated without status.".format(engine_idx))
+            job.error(f"engine_{engine_idx}: Engine terminated without status.")
 
         if mode == "bmc" or mode == "cover":
             job.update_status(task_status)
             task_status_lower = task_status.lower() if task_status == "PASS" else task_status
-            job.log("engine_{}: Status returned by engine: {}".format(engine_idx, task_status_lower))
-            job.summary.append("engine_{} ({}) returned {}".format(engine_idx, " ".join(engine), task_status_lower))
+            job.log("engine_{engine_idx}: Status returned by engine: {task_status_lower}")
```
Comment (piegamesde) [inline sbysrc/sby_engine_smtbmc.py:190]: Missing f

**Review by piegamesde — COMMENTED** [review]
There cannot be enough eyes looking at the f-string refactor, so many sneaky mistakes. I think another pass only checking that all relevant strings start with `f` will be necessary. Warning: this is an extraordinarily dull task.

**Comment by clairexen** [comment]
**Regarding "Allow to set a working directory even when having multiple tasks":**

I think this is not the best way to do this, because with this the taskname will be appended only if there are multiple tasks, meaning that scripts driving sby might have a harder time predicting the resulting workdir name.

I suggest that instead `%` is replaced with the taskname, and that we require `%` in the workdir string when more than one task is specified.

**Comment by nakengelhardt** [comment]
> There cannot be enough eyes looking at the f-string refactor, so many sneaky mistakes. I think another pass only checking that all relevant strings start with `f` will be necessary. Warning: this is an extraordinarily dull task.

As the person who did the `%`-to-`.format()`, I empathize very much. Thanks for doing this! I'll try to give it a once-over as well.

We discussed type hints but none of us is in the habit of using them. Adding them when nobody will check that they stay up to date is probably worse than not having them in the first place, so we'd prefer if you removed that commit.

**Comment by piegamesde** [comment]
@clairexen How about making this different command line flags? One to set the working directory when there is no task or one to generally set the prefix. That way it would fail early if they got confused.

@nakengelhardt the type system thing is unfortunate. I added it because it helped me better understand the code base and it helped me catch mistakes early on during refactoring (I can't test all relevant code paths myself, but type checks can easily make sure I didn't get anything fundamentally wrong regardless). I can add a type checking action to the CI, that will help against rotting.

**Comment by clairexen** [comment]
> @clairexen How about making this different command line flags? One to set the working directory when there is no task or one to generally set the prefix. That way it would fail early if they got confused.

Fine with me. However, as -d right now, the first one should work with no task *or exactly one task*.

**Comment by clairexen** [comment]
> I can add a type checking action to the CI, that will help against rotting.

Please factor the type check out into its own PR so we can merge the rest and have a separate discussion about type tags in the new PR.

**Comment by piegamesde** [comment]
ping

**Pull request merged**